specs/AllowGoBackForward.md

```diff
@@ -0,0 +1,142 @@
+## C++
+
+```cpp
+wil::com_ptr<ICoreWebView2> m_webview;
+void SettingsComponent::ToggleAllowGoBack()
+{
+    // Get webView's current settings.
+    wil::com_ptr<ICoreWebView2Settings> webView2Settings;
+    CHECK_FAILURE(m_webview->get_Settings(&webView2Settings);
+
+    BOOL allowGoBack;
+    CHECK_FAILURE(webView2Settings->get_AllowGoBack(&allowGoBack));
+    if (allowGoBack)
+    {
+        CHECK_FAILURE(webView2Settings->put_AllowGoBack(FALSE));
+    }
+    else
+    {
+        CHECK_FAILURE(webView2Settings->put_AllowGoBack(TRUE));
+    }
+}
+
+void SettingsComponent::ToggleAllowGoForward()
+{
+    // Get webView's current settings.
+    wil::com_ptr<ICoreWebView2Settings> webView2Settings;
+    CHECK_FAILURE(m_webview->get_Settings(&webView2Settings);
+
+    BOOL allowGoForward;
+    CHECK_FAILURE(webView2Settings->get_AllowGoForward(&allowGoForward));
+    if (allowGoForward)
+    {
+        CHECK_FAILURE(webView2Settings->put_AllowGoForward(FALSE));
+    }
+    else
+    {
+        CHECK_FAILURE(webView2Settings->put_AllowGoForward(TRUE));
+    }
+}
+```
+
+## C#
+```c#
+private WebView2 _webView;
+void ToggleAllowGoBack(object target, ExecutedRoutedEventArgs e)
+{
+    var webView2Settings = _webView.CoreWebView2.Settings;
+    if (webView2Settings.AllowGoBack)
+    {
+        webView2Settings.AllowGoBack = false;
+    }
+    else
+    {
+        webView2Settings.AllowGoBack = true;
+    }
+}
+void ToggleAllowGoForward(object target, ExecutedRoutedEventArgs e)
+{
+    var webView2Settings = _webView.CoreWebView2.Settings;
+    if (webView2Settings.AllowGoForward)
+    {
+        webView2Settings.AllowGoForward = false;
+    }
+    else
+    {
+        webView2Settings.AllowGoForward = true;
+    }
+}
+```
+
+# Remarks
+
+# API Notes
+See [API Details](#api-details) section below for API reference.
+
+# API Details
+
+## Win32 C++
+```c++
+/// A continuation of the ICoreWebView2Settings interface to manage allowing navigating
+/// back and forward.
+[uuid(d21886f5-03fd-4029-8179-d7ed9f726b06), object, pointer_default(unique)]
+interface ICoreWebView2Settings : IUnknown {
+  /// `AllowGoBack` controls whether allow user to navigate back. If it is false, we
+  /// will block all back navigations regardless of source (button click, xbutton click,
+  /// key event and etc.).
+  /// The `CanGoBack` API will also return false if we set `AllowGoBack` to disabled.
+  /// This setting changes will work in the next navigation, which means that changes made
+  /// after `NavigatingStarting` event do not apply until the next top-level navigation.
+  /// The default value is true.
+  [propget] HRESULT AllowGoBack([out, retval] BOOL* allow);
+  /// Set the `AllowGoBack` property.
+  [propput] HRESULT AllowGoBack([in] BOOL allow);
+  /// `AllowGoForward` controls whether allow user to navigate forward. If it is false, we
+  /// will block all forward navigations regardless of source (button click, xbutton click,
+  /// key event and etc.).
```

Comment:
You only mention end-user-controlled UI that is blocked. Does this also apply to the DOM mechanisms for controlling back/forward navigation? For example [history.back](https://developer.mozilla.org/en-US/docs/Web/API/History/back), history.go(-1). Or with [history.pushState](https://developer.mozilla.org/en-US/docs/Web/API/History/pushState) which adds a kind of virtual navigation that the HTML document can handle on its own in JavaScript but there's no actual network or page navigation. Are these sorts of things blocked as well?

Or if this whole feature is really about the end user navigating via keyboard/mouse, we can find a different name for this setting to make it clear this is only about disabling user input navigations.

Reply:
Thanks for your reminding very much! I didn't realize that we can use DOM interfaces to navigate back/forward/anywhere. By now it cannot block this kind of navigations because `history.back/forward/go` will call `NavigationControllerImpl::GoToOffsetFromRenderer` directly and will not check `NavigationControllerImpl::CanGoBack`. I'm not sure whether we need to block the DOM interfaces(seems need in my opinion). @champnic Could you help confirm do we need to block it? If need, I will investigate more on how to implement it and update this API spec document, otherwise we can find a different name for the setting. Thanks a lot!

Reply:
We could name this Settings.IsUserInputNavigationEnabled or something? There's an existing [Settings.IsSwipeNavigationEnabled](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2settings6?view=webview2-1.0.1108.44#get_isswipenavigationenabled)

Reply:
@liulichen0603 I would revisit the GitHub threads and potentially ask them if they need this support (block history.back) or not.

Reply:
Hi @champnic , thanks for helping ask them! Not sure is there any feedback from them?

Reply:
@liulichen0603 Oh sorry, I meant that I think you should follow up with the developers yourself. Typically once we begin designing an API the devs or PMs who own the feature will reach out to get more info as needed.

Reply:
Oh, sorry for my misunderstanding! I'll try to reach out on the GitHub threads and thanks for your reminding.

Reply:
Hi @david-risney @champnic , I revisited the GubHub threads and got the feedback from them:

In my cases, the DOM mechanisms don't really matters. Our final users never have a chance to execute history.back or history.go(-1). But in my eyes, it should be better to block it too.
You could look for others' opinions and make the final decision. :)

Then synced with our PM and believed that it would be better to block the DOM mechanisms. I suppose we can change the APIs to something like below:
Settings.IsUserInputNavigationEnabled { get; set; }
Settings.IsDOMNavigationEnabled { get; set; }

Do you have any concerns or suggestions about it?

Reply:
Hm. I'm not sure what the use case would be for blocking the history usage. I think that API is required for some websites to work properly. And if the website is built by the app developer, then they can just not call that API. If the use case is to display a webpage and not allow leaving the webpage at all, then we wouldn't want to just block history, but all navigations. If a page can't call history.back(), they could just directly navigate somewhere else instead.

My leaning would be `Settings.IsBackForwardEnabled` and limit it to user interaction. 

We have existing APIs that are meant to block the user from doing something, but still allow the APIs to programmatically invoke the feature, like Settings.IsDevToolsEnabled (only blocks the end-user from opening devtools - doesn't block the OpenDevTools API).